### PR TITLE
Friendship feature

### PR DESCRIPTION
**Pull request by janegreene**
Added a friendship join table, friends are a self-referential to users. 
created methods for inverse friendships and tested throughly
created a partial for flash message handling
created logic in dashboard view to add and display friends

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,6 +1,10 @@
 class User < ApplicationRecord
   validates :name, :email, :access_token, presence: true
   validates :uid, uniqueness: true, presence: true
+  has_many :friendships, dependent: :destroy
+  has_many :friends, through: :friendships
+  # has_many :inverse_friendships, class_name: 'Friendship', foreign_key: 'friend_id', dependent: :destroy
```
Comment (janegreene) [inline app/models/user.rb:6]: I need to remove these comments.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -4,4 +4,9 @@ def create
     session[:user_id] = user.id
     redirect_to '/dashboard'
   end
+
+  def destroy
+    session.delete(:user_id)
+    redirect_to '/'
+  end
 end
```
Comment (michaeljevans) [inline app/controllers/sessions_controller.rb:12]: i dig

File: app/models/friendship.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def find_inverse
+    self.class.where(inverse_match_options)
+  end
+
+  def inverse_match_options
+    { friend_id: user_id, user_id: friend_id }
+  end
+end
```
Comment (michaeljevans) [inline app/models/friendship.rb:27]: these methods are spicy

File: app/views/dashboard/show.html.erb [hunk 4]
```diff
@@ -6,7 +6,20 @@
...
 <section class='friends'>
   <p>Friends</p>
+  <%= form_tag '/friendships', method: :post %>
+  <%= label_tag :friend_email, "Friend's Email" %>
+  <%= text_field_tag :friend_email %>
+  <%= submit_tag "Add Friend" %>
+  <%= render partial: '/partials/flash' %>
   <hr>
+  <% if current_user.friends.empty?  %>
+    <p>You currently have no friends</p>
```
Comment (michaeljevans) [inline app/views/dashboard/show.html.erb:16]: :(

File: app/views/partials/_flash.html.erb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<% flash.each do |name, message| %>
+  <%= message %>
+<% end %>
```
Comment (michaeljevans) [inline app/views/partials/_flash.html.erb:3]: this will come in handy

File: spec/features/user/friendship_spec.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+      within ".friends" do
+        fill_in "Email", with: 'junk@gmail.com'
+        click_button "Add Friend"
+      end
+      expect(page).to have_content("Sorry your friend cannot be found.")
+    end
+  end
+
+end
```
Comment (michaeljevans) [inline spec/features/user/friendship_spec.rb:49]: what a test

File: spec/models/friendship_spec.rb [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    it ".find_inverse" do
+      friendship_2 = Friendship.last
+      expect(@friendship.find_inverse.first).to eq(friendship_2)
+    end
+    it ".inverse_match_options" do
+      expect(@friendship.inverse_match_options).to eq({friend_id: @user.id, user_id: @user_2.id})
+    end
+
+  end
+end
```
Comment (michaeljevans) [inline spec/models/friendship_spec.rb:47]: very thorough

**Pull request merged**